Using what you learned today, design a MLP model to train the Fashion MNIST data for the best accuracy in test. You can use either sequential API or functional API. Describe how you come to your choice of hyperparameters. What’s the best result you can achieve?

In [ ]:
import sys
import tensorflow as tf
assert tf.__version__ >= "2.0"
import numpy as np
import os
import matplotlib.pyplot as plt

from tensorflow import keras

from matplotlib import pyplot
from keras.datasets import fashion_mnist

In [12]:
(X_train_full, y_train_full), (X_test, y_test) = fashion_mnist.load_data()

In [13]:
X_valid, X_train = X_train_full[:5000] / 255., X_train_full[5000:] / 255.
y_valid, y_train = y_train_full[:5000], y_train_full[5000:]
X_test = X_test / 255.

In [14]:
print(X_valid.shape)
print(X_train.shape)
print(y_valid.shape)
print(y_train.shape)
print(X_test.shape)

(5000, 28, 28)
(55000, 28, 28)
(5000,)
(55000,)
(10000, 28, 28)


In [15]:
class_names = ["T-shirt/top", "Trouser", "Pullover", "Dress", "Coat",
               "Sandal", "Shirt", "Sneaker", "Bag", "Ankle boot"]

# Epochs

In [66]:
model = keras.models.Sequential([
    keras.layers.Flatten(input_shape=[28, 28]),
    keras.layers.Dense(300, activation="relu"),
    keras.layers.Dense(100, activation="relu"),
    keras.layers.Dense(10, activation="softmax")
])

In [67]:
model.compile(loss="sparse_categorical_crossentropy",
              optimizer="sgd",
              metrics=["accuracy"])

In [68]:
history = model.fit(X_train, y_train, epochs=30,
                    validation_data=(X_valid, y_valid))

Epoch 1/30
1719/1719 [==============================] - 7s 4ms/step - loss: 0.7006 - accuracy: 0.7723 - val_loss: 0.5055 - val_accuracy: 0.8302
Epoch 2/30
1719/1719 [==============================] - 6s 4ms/step - loss: 0.4837 - accuracy: 0.8314 - val_loss: 0.4490 - val_accuracy: 0.8422
Epoch 3/30
1719/1719 [==============================] - 6s 4ms/step - loss: 0.4410 - accuracy: 0.8462 - val_loss: 0.4459 - val_accuracy: 0.8414
Epoch 4/30
1719/1719 [==============================] - 6s 4ms/step - loss: 0.4132 - accuracy: 0.8546 - val_loss: 0.4049 - val_accuracy: 0.8566
Epoch 5/30
1719/1719 [==============================] - 6s 4ms/step - loss: 0.3950 - accuracy: 0.8609 - val_loss: 0.3952 - val_accuracy: 0.8604
Epoch 6/30
1719/1719 [==============================] - 6s 4ms/step - loss: 0.3790 - accuracy: 0.8664 - val_loss: 0.4043 - val_accuracy: 0.8594
Epoch 7/30
1719/1719 [==============================] - 6s 4ms/step - loss: 0.3675 - accuracy: 0.8695 - val_loss: 0.3635 - val_accuracy:

In [69]:
score_sgd = model.evaluate(X_test, y_test)

313/313 [==============================] - 1s 2ms/step - loss: 0.3365 - accuracy: 0.8780


In [70]:
history = model.fit(X_train, y_train, epochs=25,
                    validation_data=(X_valid, y_valid))

Epoch 1/25
1719/1719 [==============================] - 6s 4ms/step - loss: 0.2260 - accuracy: 0.9174 - val_loss: 0.2945 - val_accuracy: 0.8914
Epoch 2/25
1719/1719 [==============================] - 6s 4ms/step - loss: 0.2224 - accuracy: 0.9190 - val_loss: 0.2933 - val_accuracy: 0.8922
Epoch 3/25
1719/1719 [==============================] - 6s 4ms/step - loss: 0.2184 - accuracy: 0.9211 - val_loss: 0.3125 - val_accuracy: 0.8908
Epoch 4/25
1719/1719 [==============================] - 6s 4ms/step - loss: 0.2143 - accuracy: 0.9224 - val_loss: 0.2905 - val_accuracy: 0.8968
Epoch 5/25
1719/1719 [==============================] - 6s 4ms/step - loss: 0.2111 - accuracy: 0.9238 - val_loss: 0.3402 - val_accuracy: 0.8760
Epoch 6/25
1719/1719 [==============================] - 6s 4ms/step - loss: 0.2087 - accuracy: 0.9253 - val_loss: 0.3215 - val_accuracy: 0.8870
Epoch 7/25
1719/1719 [==============================] - 6s 4ms/step - loss: 0.2044 - accuracy: 0.9273 - val_loss: 0.2940 - val_accuracy:

In [71]:
score = model.evaluate(X_test, y_test)

313/313 [==============================] - 1s 2ms/step - loss: 0.3425 - accuracy: 0.8839


In [72]:
history = model.fit(X_train, y_train, epochs=45,
                    validation_data=(X_valid, y_valid))

Epoch 1/45
1719/1719 [==============================] - 6s 4ms/step - loss: 0.1523 - accuracy: 0.9461 - val_loss: 0.3180 - val_accuracy: 0.8926
Epoch 2/45
1719/1719 [==============================] - 6s 4ms/step - loss: 0.1490 - accuracy: 0.9475 - val_loss: 0.2969 - val_accuracy: 0.9016
Epoch 3/45
1719/1719 [==============================] - 6s 4ms/step - loss: 0.1478 - accuracy: 0.9473 - val_loss: 0.2886 - val_accuracy: 0.8996
Epoch 4/45
1719/1719 [==============================] - 6s 4ms/step - loss: 0.1446 - accuracy: 0.9487 - val_loss: 0.3120 - val_accuracy: 0.8926
Epoch 5/45
1719/1719 [==============================] - 6s 4ms/step - loss: 0.1420 - accuracy: 0.9499 - val_loss: 0.3014 - val_accuracy: 0.9016
Epoch 6/45
1719/1719 [==============================] - 6s 4ms/step - loss: 0.1403 - accuracy: 0.9504 - val_loss: 0.3063 - val_accuracy: 0.8964
Epoch 7/45
1719/1719 [==============================] - 6s 4ms/step - loss: 0.1380 - accuracy: 0.9502 - val_loss: 0.2963 - val_accuracy:

In [73]:
score = model.evaluate(X_test, y_test)

313/313 [==============================] - 1s 2ms/step - loss: 0.4027 - accuracy: 0.8883


In [74]:
input_ = keras.layers.Input(shape=X_train.shape[1:])

hidden1 = keras.layers.Dense(30, activation="relu")(input_)
hidden2 = keras.layers.Dense(30, activation="relu")(hidden1) 

concat = keras.layers.concatenate([input_, hidden2])
output = keras.layers.Dense(1)(concat)
model = keras.models.Model(inputs=[input_], outputs=[output])

model.summary()

Model: "model"
__________________________________________________________________________________________________
 Layer (type)                   Output Shape         Param #     Connected to                     
 input_1 (InputLayer)           [(None, 28, 28)]     0           []                               
                                                                                                  
 dense_13 (Dense)               (None, 28, 30)       870         ['input_1[0][0]']                
                                                                                                  
 dense_14 (Dense)               (None, 28, 30)       930         ['dense_13[0][0]']               
                                                                                                  
 concatenate (Concatenate)      (None, 28, 58)       0           ['input_1[0][0]',                
                                                                  'dense_14[0][0]']           

# Layers

In [94]:
model = keras.models.Sequential([
    keras.layers.Flatten(input_shape=[28, 28]),
    keras.layers.Dense(350, activation="relu"),
     keras.layers.Dense(300, activation="relu"),
    keras.layers.Dense(100, activation="relu"),
    keras.layers.Dense(10, activation="softmax")
])

In [95]:
model.compile(loss="sparse_categorical_crossentropy",
              optimizer="sgd",
              metrics=["accuracy"])

In [96]:
history = model.fit(X_train, y_train, epochs=30,
                    validation_data=(X_valid, y_valid))

Epoch 1/30
1719/1719 [==============================] - 11s 6ms/step - loss: 0.7271 - accuracy: 0.7574 - val_loss: 0.4750 - val_accuracy: 0.8378
Epoch 2/30
1719/1719 [==============================] - 10s 6ms/step - loss: 0.4768 - accuracy: 0.8337 - val_loss: 0.4229 - val_accuracy: 0.8554
Epoch 3/30
1719/1719 [==============================] - 10s 6ms/step - loss: 0.4282 - accuracy: 0.8497 - val_loss: 0.4150 - val_accuracy: 0.8560
Epoch 4/30
1719/1719 [==============================] - 10s 6ms/step - loss: 0.4006 - accuracy: 0.8579 - val_loss: 0.3780 - val_accuracy: 0.8696
Epoch 5/30
1719/1719 [==============================] - 10s 6ms/step - loss: 0.3793 - accuracy: 0.8651 - val_loss: 0.3707 - val_accuracy: 0.8702
Epoch 6/30
1719/1719 [==============================] - 10s 6ms/step - loss: 0.3613 - accuracy: 0.8723 - val_loss: 0.3672 - val_accuracy: 0.8696
Epoch 7/30
1719/1719 [==============================] - 10s 6ms/step - loss: 0.3473 - accuracy: 0.8759 - val_loss: 0.3602 - val_ac

In [97]:
score = model.evaluate(X_test, y_test)

313/313 [==============================] - 1s 3ms/step - loss: 0.3312 - accuracy: 0.8840


In [98]:
model = keras.models.Sequential([
    keras.layers.Flatten(input_shape=[28, 28]),
    keras.layers.Dense(500, activation="relu"),
     keras.layers.Dense(350, activation="relu"),
    keras.layers.Dense(300, activation="relu"),
    keras.layers.Dense(100, activation="relu"),
    keras.layers.Dense(10, activation="softmax")
])

In [99]:
model.compile(loss="sparse_categorical_crossentropy",
              optimizer="sgd",
              metrics=["accuracy"])

In [100]:
history = model.fit(X_train, y_train, epochs=30,
                    validation_data=(X_valid, y_valid))

Epoch 1/30
1719/1719 [==============================] - 36s 9ms/step - loss: 0.7068 - accuracy: 0.7630 - val_loss: 0.5278 - val_accuracy: 0.8180
Epoch 2/30
1719/1719 [==============================] - 15s 9ms/step - loss: 0.4627 - accuracy: 0.8364 - val_loss: 0.4654 - val_accuracy: 0.8302
Epoch 3/30
1719/1719 [==============================] - 15s 9ms/step - loss: 0.4098 - accuracy: 0.8545 - val_loss: 0.3863 - val_accuracy: 0.8614
Epoch 4/30
1719/1719 [==============================] - 15s 8ms/step - loss: 0.3785 - accuracy: 0.8646 - val_loss: 0.3620 - val_accuracy: 0.8694
Epoch 5/30
1719/1719 [==============================] - 15s 9ms/step - loss: 0.3570 - accuracy: 0.8703 - val_loss: 0.3796 - val_accuracy: 0.8636
Epoch 6/30
1719/1719 [==============================] - 15s 9ms/step - loss: 0.3396 - accuracy: 0.8763 - val_loss: 0.3398 - val_accuracy: 0.8788
Epoch 7/30
1719/1719 [==============================] - 15s 9ms/step - loss: 0.3238 - accuracy: 0.8821 - val_loss: 0.3269 - val_ac

In [101]:
score = model.evaluate(X_test, y_test)

313/313 [==============================] - 1s 4ms/step - loss: 0.3216 - accuracy: 0.8909


# Optimizers
1.Adagrad
2.Adamax
3.RMSprop
4.SGD

In [82]:
model.compile(loss="sparse_categorical_crossentropy",
              optimizer="Adagrad",
              metrics=["accuracy"])

In [83]:
history = model.fit(X_train, y_train, epochs=30,
                    validation_data=(X_valid, y_valid))

Epoch 1/30
1719/1719 [==============================] - 17s 10ms/step - loss: 0.1216 - accuracy: 0.9575 - val_loss: 0.2885 - val_accuracy: 0.9024
Epoch 2/30
1719/1719 [==============================] - 17s 10ms/step - loss: 0.1144 - accuracy: 0.9606 - val_loss: 0.2846 - val_accuracy: 0.9054
Epoch 3/30
1719/1719 [==============================] - 18s 10ms/step - loss: 0.1110 - accuracy: 0.9619 - val_loss: 0.2883 - val_accuracy: 0.9074
Epoch 4/30
1719/1719 [==============================] - 18s 10ms/step - loss: 0.1079 - accuracy: 0.9638 - val_loss: 0.2904 - val_accuracy: 0.9048
Epoch 5/30
1719/1719 [==============================] - 19s 11ms/step - loss: 0.1054 - accuracy: 0.9644 - val_loss: 0.2942 - val_accuracy: 0.9042
Epoch 6/30
1719/1719 [==============================] - 18s 11ms/step - loss: 0.1035 - accuracy: 0.9650 - val_loss: 0.2888 - val_accuracy: 0.9048
Epoch 7/30
1719/1719 [==============================] - 18s 11ms/step - loss: 0.1014 - accuracy: 0.9662 - val_loss: 0.2898 -

In [84]:
score_adagrad = model.evaluate(X_test, y_test)

313/313 [==============================] - 1s 4ms/step - loss: 0.3393 - accuracy: 0.8952


In [85]:
model.compile(loss="sparse_categorical_crossentropy",
              optimizer="Adamax",
              metrics=["accuracy"])

In [ ]:
history = model.fit(X_train, y_train, epochs=30,
                    validation_data=(X_valid, y_valid))

Epoch 1/30
1719/1719 [==============================] - 18s 10ms/step - loss: 0.2461 - accuracy: 0.9102 - val_loss: 0.3439 - val_accuracy: 0.8802
Epoch 2/30
1719/1719 [==============================] - 17s 10ms/step - loss: 0.2258 - accuracy: 0.9143 - val_loss: 0.3236 - val_accuracy: 0.8860
Epoch 3/30
1719/1719 [==============================] - 18s 10ms/step - loss: 0.2130 - accuracy: 0.9204 - val_loss: 0.3201 - val_accuracy: 0.8870
Epoch 4/30
1719/1719 [==============================] - 18s 10ms/step - loss: 0.1992 - accuracy: 0.9238 - val_loss: 0.3331 - val_accuracy: 0.8894
Epoch 5/30
1719/1719 [==============================] - 17s 10ms/step - loss: 0.1890 - accuracy: 0.9281 - val_loss: 0.3016 - val_accuracy: 0.9000
Epoch 6/30
1169/1719 [===================>..........] - ETA: 5s - loss: 0.1759 - accuracy: 0.9321

In [ ]:
score_adamax = model.evaluate(X_test, y_test)

In [ ]:
model.compile(loss="sparse_categorical_crossentropy",
              optimizer="RMSprop",
              metrics=["accuracy"])

In [89]:
history = model.fit(X_train, y_train, epochs=30,
                    validation_data=(X_valid, y_valid))

Epoch 1/30
1719/1719 [==============================] - 22s 13ms/step - loss: 0.4501 - accuracy: 0.8518 - val_loss: 0.4704 - val_accuracy: 0.8740
Epoch 2/30
1719/1719 [==============================] - 21s 12ms/step - loss: 0.4185 - accuracy: 0.8636 - val_loss: 0.5270 - val_accuracy: 0.8486
Epoch 3/30
1719/1719 [==============================] - 21s 12ms/step - loss: 0.4076 - accuracy: 0.8677 - val_loss: 0.5096 - val_accuracy: 0.8696
Epoch 4/30
1719/1719 [==============================] - 21s 12ms/step - loss: 0.4129 - accuracy: 0.8694 - val_loss: 0.4781 - val_accuracy: 0.8640
Epoch 5/30
1719/1719 [==============================] - 21s 12ms/step - loss: 0.4164 - accuracy: 0.8701 - val_loss: 0.6163 - val_accuracy: 0.8464
Epoch 6/30
1719/1719 [==============================] - 21s 12ms/step - loss: 0.4155 - accuracy: 0.8731 - val_loss: 0.5773 - val_accuracy: 0.8638
Epoch 7/30
1719/1719 [==============================] - 21s 12ms/step - loss: 0.4133 - accuracy: 0.8733 - val_loss: 0.6619 -

In [90]:
score_rms = model.evaluate(X_test, y_test)

313/313 [==============================] - 1s 4ms/step - loss: 1.8638 - accuracy: 0.6754


In [91]:
print("Acuuracy with SGD:",score_sgd [1])
print("Acuuracy with Adagrad:",score_adagrad [1])
print("Acuuracy with Adamax:",score_adamax [1])
print("Acuuracy with RMSprop:",score_rms [1])

Acuuracy with SGD: 0.878000020980835
Acuuracy with Adagrad: 0.8952000141143799
Acuuracy with Adamax: 0.8913999795913696
Acuuracy with RMSprop: 0.6754000186920166


# Prediction

In [92]:
X_new = X_test[:3]
y_proba = model.predict(X_new)
y_proba.round(2)

array([[0., 0., 0., 0., 0., 0., 0., 0., 0., 1.],
       [0., 0., 1., 0., 0., 0., 0., 0., 0., 0.],
       [0., 1., 0., 0., 0., 0., 0., 0., 0., 0.]], dtype=float32)

In [93]:
y_new = y_test[:3]
y_new

array([9, 2, 1], dtype=uint8)

# Conclusion

Epochs:
  epochs=30 - accuracy: 0.8780
  epochs=25 - accuracy: 0.8839
  epochs=45 - accuracy: 0.8883

Layers/Hyperparameter:
5 hidden layers: accuracy - 0.8840
6 hidden layers: accuracy - 0.8909
As the number of hidden layer increase the accuracy increases.

Optimizers:
  SGD - 0.8780
  Adagrad - 0.8952
  Adamax - 0.8913
  RMSprop - 0.6754
  Out of the 4 optimizers adagrad is the most accurate